### PR TITLE
fix: WaitOnSocket select error when sockfd above FD_SETSIZE

### PR DESCRIPTION
**Pull request by taozhijiang**
Fixes select return code is not well handled, and when curl sockfd above FD_SETSIZE (1024), select can not handle it.

Changes:
1. According docs, select return -1 on error, 0 when timeout, and >0 for expected events return. The original cast the int code to bool, so when invalid sock provided, the procedure treat as good, the following procedure crash. So the code carefully check the select return code, and the really events we get.
2. Even though select is POSIX standard and cross platform, but can not handle the case when fd above FD_SETSIZE. In our production environment, we see CURL sock fd far more above it, so when Linux environment possible, we use poll instead of select.

### CODE REVIEW

**Comment by owent** [comment]
Could you please do this change base on [async-changes](https://github.com/open-telemetry/opentelemetry-cpp/tree/async-changes) branch? The `select` call wil be removed after we merge this branch back into `main`.

**Comment by lalitb** [comment]
> Could you please do this change base on [async-changes](https://github.com/open-telemetry/opentelemetry-cpp/tree/async-changes) branch? The `select` call wil be removed after we merge this branch back into `main`.

@owent - Are these changes relevant once `async-changes` branch is merged to `main`? If not, we Should we can let these changed merged in main branch for now, and later these changes would be removed once `async-changes` branch is merged to `main`. Would leave it to you to decide :)

**Comment by lalitb** [comment]
> Could you please do this change base on [async-changes](https://github.com/open-telemetry/opentelemetry-cpp/tree/async-changes) branch? The `select` call wil be removed after we merge this branch back into `main`.

@owent - Are these changes relevant once `async-changes` branch is merged to `main`? If not, we Should we can let these changed merged in main branch for now, and later these changes would be removed once `async-changes` branch is merged to `main`. Would leave it to you to decide :)

**Comment by taozhijiang** [comment]
> > Could you please do this change base on [async-changes](https://github.com/open-telemetry/opentelemetry-cpp/tree/async-changes) branch? The `select` call wil be removed after we merge this branch back into `main`.
> 
> @owent - Are these changes relevant once `async-changes` branch is merged to `main`? If not, we Should we can let these changed merged in main branch for now, and later these changes would be removed once `async-changes` branch is merged to `main`. Would leave it to you to decide :)


This is a definite code defect which may affect the delopyment in production, so I think should be fixed ASAP.
We will try [async-changes](https://github.com/open-telemetry/opentelemetry-cpp/tree/async-changes) latter in test environment when we are available ;-)

**Comment by owent** [comment]
> relevant

Yes, the implement in `async-changes` branch will use [curl_multi_poll][1] to replace `select`. This problem may still exist because without event loop system, [curl_multi_poll][1] use `select` on some platform, but we will reuse connections for multiple sessions now, and when `max_concurrent_requests / max_requests_per_connection` is less than 1023, we will have 1024 fds at most at the same time.The default value of `max_concurrent_requests` is 64 and default value `max_requests_per_connection` is 8 now, which means will have no more than 9 connections at most by default.

[1]: https://curl.se/libcurl/c/curl_multi_poll.html

**Comment by owent** [comment]
> > > Could you please do this change base on [async-changes](https://github.com/open-telemetry/opentelemetry-cpp/tree/async-changes) branch? The `select` call wil be removed after we merge this branch back into `main`.
> > 
> > 
> > @owent - Are these changes relevant once `async-changes` branch is merged to `main`? If not, we can let these changes merged in main branch for now, and later these changes would be removed once `async-changes` branch is merged to `main`. Would leave it to you to decide :)
> 
> This is a definite code defect which may affect the delopyment in production, so I think should be fixed ASAP. We will try [async-changes](https://github.com/open-telemetry/opentelemetry-cpp/tree/async-changes) latter in test environment when we are available ;-)

Sure, if you don't mind these changes in this PR may be removed later.
This PR also fixes #1220 .

File: ext/include/opentelemetry/ext/http/client/curl/http_operation_curl.h
Comment (owent) [inline ext/include/opentelemetry/ext/http/client/curl/http_operation_curl.h:484]: Some compiler will warning this(implicit convert from int to bool).We should use `res = 0 != FD_ISSET(sockfd, &infd);` here.

File: ext/include/opentelemetry/ext/http/client/curl/http_operation_curl.h
Comment (owent) [inline ext/include/opentelemetry/ext/http/client/curl/http_operation_curl.h:455]: Should we retry several times when the `sockfd` is not selected? It usually happens when there is a lot actived fds, and `select` can not get them out once.

**Review by owent — CHANGES_REQUESTED** [review]

File: ext/include/opentelemetry/ext/http/client/curl/http_operation_curl.h
Comment (taozhijiang) [reply ext/include/opentelemetry/ext/http/client/curl/http_operation_curl.h:455]: Our service is quit busy.
Even though I set the trace sample_ration=0.01, the service startup several minutes latter, the CURLINFO_LASTSOCKET return sock fd above FD_SETSIZE. After this, all the trace message can not be upload and never recovered.
I don't think retry can work.

**Review by owent — COMMENTED** [review]
LGTM after the format is fixed.

**Comment by lalitb** [comment]
@taozhijiang  - You can fix the format as documented here - https://github.com/open-telemetry/opentelemetry-cpp/blob/main/CONTRIBUTING.md#pull-requests

**Review by lalitb — APPROVED** [review]

**Comment by lalitb** [comment]
> LGTM after the format is fixed.

@owent  - Can you remove the requested change if it looks good to merge? Thanks.

**Review by owent — APPROVED** [review]

**Comment by owent** [comment]
> > LGTM after the format is fixed.
> 
> @owent - Can you remove the requested change if it looks good to merge. Thanks.

Done

**Pull request merged**
